# Ujian Akhir Semester (3)

Nama : Clarence Chandra Sentosa

NRP  : 1973010

# Training Classifier dengan Naive Bayes

# Load Breast Cancer Dataset

In [1]:
# Code 01

from sklearn.datasets import load_breast_cancer

print(load_breast_cancer().DESCR)

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        worst/largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 0 is Mean Radi

**Konsep berpikir code 01 :**

- Pertama-tama kita akan mengimport terlebih dahulu modulnya dengan memanggil method "from sklearn.datasets import load_breast_cancer".
- Lalu selajutnya kita akan memanggil "load_breast_cancer()" dan disini kita akan coba akses description dari datasetnya dengan memanggil ".DESCR".
- Lalu terakhir kita akan coba tampilkna pada layar dengan memanfaatkan fungsi print.

Berdasarkan hasil code 01, kita bisa melihat isi deskripsi dari dataset "Breast Cancer Wisconsin (Diagnostic)", kita bisa ketahui juga bahwa dataset ini terdiri dari 569 instances atau 569 data, kita juga bisa lihat number of attributes adalah 30 numeric yang artinya disini terdapat 30 nilai numerik sebagai featurenya atau sebagai kolomnya yang kita bisa lihat featurenya terdiri dari : - radius (mean of distances from center to points on the perimeter) - texture (standard deviation of gray-scale values) - perimeter - area - smoothness (local variation in radius lengths) - compactness (perimeter^2 / area - 1.0) - concavity (severity of concave portions of the contour) - concave points (number of concave portions of the contour) - symmetry - fractal dimension ("coastline approximation" - 1)

Semua feature tersebut memiliki type data numeric dengan total jumlah feature adalah 30. Terkait dengan class, terdapat 2 buah class pada dataset ini yaitu WDBC-Malignant (mengindikasikan tumor ganas) dan WDBC-Benign (mengindikasikan tumor jinak). Kita juga bisa melihat bahwa pada dataset ini tidak terdapat missing value. Lalu terkait proporsi classnya, terdapat 212 data yang diklasifikasikan sebagai malignant atau tumor ganas dan terdapat 357 data yang diklasifikasikan sebagai benign atau tumor jinak. Lalu selanjutnya ada keterangan creator ata pembuat dataset, dan ada juga tanggal pembuatannya. Kita sebaiknya mengetahui deskirpisi suatu dataset terlebih dahulu sebelum bekerja dengan dataset yang akan digunakan.

# Dataset Dimension

In [2]:
# Code 02

X, y = load_breast_cancer(return_X_y=True)
X.shape

(569, 30)

**Konsep berpikir code 02 :**

- Pertama-tama kita panggil method *load_breast_cancer* dengan menyertakan parameter (return_X_y=True).
- Lalu selanjutnya kita panggil *X.shape* dan kita coba run.

Bisa kita lihat pada output code 02 bahwa, jumlah baris pada **Breast Cancer** dataset tersebut berjumlah 569 baris dan 30 kolom.

# Training & Testing Set

Kita akan membagi dataset tersebut kedalam *training set* dan *testing set* nya.

In [3]:
# Code 03

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X,
                                                    y,
                                                    test_size=0.2,
                                                    random_state=0)

print(f'X_train shape {X_train.shape}')
print(f'X_test shape {X_test.shape}')

X_train shape (455, 30)
X_test shape (114, 30)


**Konsep berpikir code 03 :**
- Pertama-tama kita akan mencoba import terlebih dahulu modulnya dengan memanggil fungsi "from sklearn.model_selection import train_test_split".
- Selanjutnya kita akan memanggil "train_test_split" dengan menyertakan parameter "(X, y, test_size=0.2 (artinya 20% dari data kita akan dialokasikan untuk testing dataset, sedangkan 80% akan digunakan sebagai training dataset), random_state=0). Fungsi ini akan mengembalikan 4 buah kumpulan nilai yang perlu kita tangkap kedalam 4 variabel yaitu "X_train, X_test, y_train, dan y_test".
- Untuk selanjutnya, kita akan coba untuk menampilkan dimensi dari X_train dan X_test nya.

Bisa kita lihat pada code 03, untuk X_train jumlah datanya ada 455 dan untuk X_test ada 114 artinya X_train merupakan 80% dan X_test 20% dari keseluruhan dataset yang kita miliki.

# Naive Bayes dengan Scikit Learn

Pada kali ini kita akan menerapkan "Naive Bayes" untuk melakukan klasifikasi data pada dataset "Breast Cancer" yang kita miliki.

In [8]:
# Code 04

from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score

model = GaussianNB()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)
f'accuracy_score : {accuracy_score(y_test, y_pred)}'

'accuracy_score : 0.9298245614035088'

**Konsep berpikir code 04 :**
- Pertama-tama kita mengimport terlebih dahulu modelnya dengan memanggil fungsi "from sklearn.naive_bayes import GaussianNB". NB merepresentasikan Naive Bayes.
- Lalu kita juga akan mengimport matriks akurasinya dengan memanggi fungsi "from sklearn.metrics import accuracy_score".
- Berikutnya kita aka membentuk terlebih dahulu objek dari model "GaussianNB()" yang ditampung kedalam variabel "model".
- Lalu selanjutnya, model ini akan training dengan menggunakan method "fit()" dengan menyertakan parameter "(X_train, y_train)".
- Setelah modelnya kita training, maka train model tersebut akan kita gunakan untuk melakukan prediksi terhadap nilai dari X_test yang hasil prediksinya akan kita tampung kedalam variabel "y_pred".
- Untuk selanjutnya, kita akan mengukur nilai akurasinya dengan memanggil fungsi "accuracy_score" dengan menyertakan parameter "(y_test, y_pred)"

Kita bisa lihat bahwa hasil akurasinya adalah 0.9298245614035088. Kita juga bisa mencari nilai akurasinya dengan cara yang lebih sederhana (perhatikan pada code 05).

In [9]:
# Code 05

model.score(X_test, y_test)

0.9298245614035088

Selanjutnya kita akan melakukan evaluasi performa dari Naive Bayes model yang baru saja kita training tadi dengan memanfaatkan Classification Report.

In [10]:
# Code 06

from sklearn.metrics import classification_report

y_pred = model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.91      0.91      0.91        47
           1       0.94      0.94      0.94        67

    accuracy                           0.93       114
   macro avg       0.93      0.93      0.93       114
weighted avg       0.93      0.93      0.93       114



**Konsep berpikir code 06 :

- Pertama-tama kita akan import terlebih dahulu modulnya dengan memanggil method "from sklearn.metrics import classification_report".
- Lalu berikutnya kita akan melakukan prediksi dengan memanggil method "model.predict" dengan memanfaatkan testing_set kita (X_test), yang hasil prediksinya akan ditampung kedalam variabel y_pred.
- Untuk selanjutnya akan kita evaluasi dengan menggunakan classification_report dengan menyertakan parameter y_test dan y_pred.
- Pada output code 06, kita bisa melihat kolom nilai precission, recall, f1-score, support, serta kolom nilai accuracy nya sebagai peforma dari Naive Bayes model yang telah kita gunakan.

Demikianlah cara yang saya lakukan untuk melakukan training classifier dengan menggunakan *Naive Bayes* model pada **Breast Cancer** dataset.

# "Banyak Belajar Biar Bisa Bantu Banyak Orang"

Written by : Clarence Code Pianist (1973010)